ga4gh/backend.py

```diff
@@ -762,6 +770,69 @@ def runSearchDatasets(self, request):
             protocol.SearchDatasetsResponse,
             self.datasetsGenerator)
 
+    def runSearchGenotypePhenotype(self, requestStr):
+        """
+        Runs the specified SearchGenotypePhenotypeRequest.
+        We do not attempt to re-use runSearchRequest as
+        All you really need is to get the offset from the pageToken, and then
+        pass this on to the query method in the data model object.
+        """
+        # TODO verify evidence maps to drug
+
+        # validate request
+        self.startProfile()
+        try:
+            requestDict = json.loads(requestStr)
+        except ValueError:
+            raise exceptions.InvalidJsonException(requestStr)
+        requestClass = protocol.SearchGenotypePhenotypeRequest
+        responseClass = protocol.SearchGenotypePhenotypeResponse
+        self.validateRequest(requestDict, requestClass)
+        request = requestClass.fromJsonDict(requestDict)
+        if request.pageSize is None:
+            request.pageSize = self._defaultPageSize
+        if request.pageSize <= 0:
+            raise exceptions.BadPageSizeException(request.pageSize)
+        responseBuilder = protocol.SearchResponseBuilder(
+            responseClass, request.pageSize, self._maxResponseLength)
+        # ensure request has data we can query
+        if (request.evidence is None and
+                request.phenotype is None and
+                request.feature is None):
+            msg = "Error:One of evidence,phenotype or feature must be non-null"
+            raise exceptions.BadRequestException(msg)
+        # determine offset for paging
+        if request.pageToken is not None:
+            request.offset = int(request.pageToken.split(':')[1])
+        else:
+            request.offset = 0
+
+        offset = request.offset
+        count = 0
+        iterator = self._g2pDataset._search(request)
+        for obj in iterator:
+            responseBuilder.addValue(obj)
+            offset += 1
+            count += 1
+            if responseBuilder.isFull():
+                break
+
+        nextPageToken = None
+        if request.pageSize == count:
+            nextPageToken = "{}:{}".format(0, offset)
+
+        responseBuilder.setNextPageToken(nextPageToken)
+        responseString = responseBuilder.getJsonString()
+        self.validateResponse(responseString, responseClass)
+        self.endProfile()
+        return responseString
+
+        # # old ......
+        # return self.runSearchRequest(
+        #     request, protocol.SearchGenotypePhenotypeRequest,
+        #     protocol.SearchGenotypePhenotypeResponse,
+        #     self.genotypePhenotypeGenerator)
+
 
```

Comment:
This is redoing quite a lot of things, and we're breaking layer a bit as well. I've played with this a bit, and here's what I suggest as a starting point:
```python
    def genotypePhenotypeGenerator(self, request):
        if (request.evidence is None and
                request.phenotype is None and
                request.feature is None):
            msg = "Error:One of evidence,phenotype or feature must be non-null"
            raise exceptions.BadRequestException(msg)
        annotationList = self._g2pDataset.queryLabels(
            request.feature, request.evidence, request.phenotype)
        return self._objectListGenerator(request, annotationList)
        
    def runSearchGenotypePhenotype(self, request):
        return self.runSearchRequest(
            request, protocol.SearchGenotypePhenotypeRequest,
            protocol.SearchGenotypePhenotypeResponse,
            self.genotypePhenotypeGenerator)
```

This implemenatation pawns off all the paging work to the ``_objectListGenerator``, which expects a list of objects that have a ``toProtocolElement`` method. If you return a list of Annotation objects from ``queryLabels`` then this should be easy.

This has major disadvantages, as we get the list of all associations every time we make a request. I wonder how much of a problem this will be in practise though. Do we expect to stream millions of annotations? Managing the paging properly without violating all sorts of layering assumptions is a massive headache. I think this approach is probably good enough for initial purposes, and it has the advantage of being correct.